caikit/runtime/client/remote_module_base.py

```diff
@@ -0,0 +1,496 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+The RemoteModuleBase is a base class that can be mutated to have the same task methods
+as a ModuleBase but submit requests to a remote runtime instead of loading locally. By 
+design this class/factory does not use any references to the original Module class.
+"""
+# Standard
+from collections import OrderedDict
+from threading import Lock
+from typing import Any, Callable, Dict, Generator, List, Type
+import copy
+import json
+import uuid
+
+# Third Party
+import grpc
+import requests
+
+# First Party
+import alog
+
+# Local
+from caikit.config.config import merge_configs
+from caikit.core.data_model import DataBase, DataStream
+from caikit.core.exceptions import error_handler
+from caikit.core.exceptions.caikit_core_exception import (
+    CaikitCoreException,
+    CaikitCoreStatusCode,
+)
+from caikit.core.modules import ModuleBase, module
+from caikit.core.task import TaskBase
+from caikit.interfaces.common.data_model.remote import ConnectionInfo
+from caikit.runtime.client.remote_config import RemoteModuleConfig, RemoteRPCDescriptor
+from caikit.runtime.names import (
+    MODEL_ID,
+    OPTIONAL_INPUTS_KEY,
+    REQUIRED_INPUTS_KEY,
+    ServiceType,
+    get_grpc_route_name,
+    get_http_route_name,
+)
+
+log = alog.use_channel("RMBASE")
+error = error_handler.get(log)
+
+
+class RemoteModelBaseClass(ModuleBase):
+    """Private class to act as the base for remote modules. This class will be subclassed and
+    mutated by construct_remote_module_class to make it have the same functions and parameters
+    as the source module."""
+
+    def __init__(
+        self,
+        connection_info: ConnectionInfo,
+        protocol: str,
+        model_key: str,
+        model_name: str,
+    ):
+        # Initialize module base
+        super().__init__()
+
+        self._model_name = model_name
+
+        # Load connection parameters
+        self._connection = connection_info
+        self._tls = self._connection.tls
+        self._protocol = protocol
+        self._model_key = model_key
+
+        # Configure GRPC variables and threading lock
+        self._channel_lock = Lock()
+        self.__grpc_channel = None
```

Comment:
If we use `requests.Session`, we might want to expand this from `__grpc_channel` to `__conn_channel` to hold the `Session` too

Reply:
Good idea! Using `requests.Session` should clean up the class and shore up the differences between grpc and http